src/cephadm/design_choices.rst

```diff
@@ -0,0 +1,165 @@
+
+
+Downsides:
+
+We're still `scp'ing` bin/cephadm from the ceph-mgr to the host which is then executed.
+A daemon on the other hand needs to run continuously on the target system. If ceph-mgr (and the containing bin/cephadm
+script) is updated, the daemon needs to be informed of this event and get updated/restarted as well.
+
+
+1.1.1 HTTP:
+___________
+
+There is a implementation by Paul Cuzner [2] that uses HTTP to serve an endpoint to the hosts metadata.
+
+Upside:
+
+* We could utilize a proper REST API to CRUD all of the daemons/disks etc.
+  Currently it feels a bit weird to call a binary from a remote host (that's another topic though)
+
+
+Downside:
+
+* This is quite some shift in architecture.
+* We still use SSH to bootstrap a node (in order to install bin/cephadm for the first time)
+  Having two distinct transport layers feels odd.
+* The current approach of delivering bin/cephadm to the host doesn't allow the use of external dependencies.
+  This means that we're stuck with the built-in HTTP server lib, which isn't meant for production purposes.
+  bin/cephadm needs to be packaged and distributed (one way or the other) for us to make use of a production ready
+  http server.
+
+Requirements:
+
+* https
+* authentication
+* documentation
+* api versioning (keep it stable)
+
+
+1.1.2 Keep bin/cephadm
+______________________
+
+Daemonizing bin/cephadm allows us to pre-fetch data and store it in some shape or form.
+
+Currently we're only interested in speeding up metadata retrieval. This could mean that we don't even
+need to introduce a new transport layer but rather transparently return cached content via the `cephadm ls`
+or the `cephadm ceph-volume ls` command.
+
+The technical details need to be figured out. (format of data storage etc.)
+
+Upside:
+
+* No new transport layer.
+  Also means no extra auth, security layer is needed
+
+Downside:
+
+* May feel a bit hacky
+* Needs locks (if a refresh is currently happening)
+
+
+1.1.3 Use k-v store
+___________________
+
+After the mgr/cephadm queried metadata from each host, we usually save it in the mon's k-v store.
+
+If each host would be allowed to write their own metadata to the store, we'd not need to explicitly call
+for a refresh.
+
+This needs research on some topics:
+
+* Is the mon's k-v store capable of accepting **many** requests (CRUD)
+* Does the auth/keyring mechanism allow to reduce access to certain areas of the mon store?
+
+
+2 Push or Pub-Sub
+-----------------
+
+This is a change of paradigm. However:
+
+Upside:
+* Events can be captured and forwarded. Instant notifications
+* This is a proper and fast implementation for such a system
+
+Downside:
+
+* This is would require a major rewrite (also how mgr/cephadm operates)
+
+
+3 Increase the worker pool size
+--------------------------------
+
+As we identified in the `Intro` section, mgr/cephadm is currently able to scrape 10 nodes at the same time.
+
```

Comment:
the exporter PR returns the metadata in 15ms (as measured with curl). Keeping the scrape time low promotes more host scrapes in the same time window offering the potential to reduce the parallelism in mgr...not increase it.

Also, you meantion the frequency of the scrape, but the scrape for list daemons is ~ 10s, ceph-volume ~7s ... the scale issue is not just how parallel we can go, but the scrape time itself determines the refresh cycle. Without offloading to the host, and performing the scrapes asynchronously, we can not address the current scale limitations that inhibit orchestrator, and undermine the dashboard (with old state data - looking at a UI that presents a view 6mins into the past is not really viable in my view)

Reply:
Please see my notes in the next 10 lines. I've listed this option for the sake of completeness and documentation. It's not an option